### ISSUE TITLE
losing one example when using minibatch

### ISSUE THREAD

**Issue by ustbxyls** [deeplearning4j/deeplearning4j#5704]
I use minibatch to train the MLP net, but when I do the  evaluation, the confusion matrix confuses me
![image](https://user-images.githubusercontent.com/4646748/41883605-08e58f20-7922-11e8-9aa0-d0f0edf70776.png)
 
this Confusion Matrix shows that  there are **99** examples in the test set, but actually the number is **100**.

so, what's the problem?

**Comment by ustbxyls** [comment]
BTW, here is my code:
`RecordReader rr = new CSVRecordReader();
        JavaRDD<String> stringData = sc.textFile(filenameTrain);
        JavaRDD<DataSet> points = stringData.map(new RecordReaderFunction(rr, 0, numOutputs)).cache();
        JavaRDD<DataSet> trainData = new RDDMiniBatches(batchSizePerWorker, points).miniBatchesJava();

        RecordReader rrTest = new CSVRecordReader();
        JavaRDD<String> stringTestData = sc.textFile(filenameTest);
        JavaRDD<DataSet> pointsTest = stringTestData.map(new RecordReaderFunction(rrTest, 0, numOutputs)).cache();
        JavaRDD<DataSet> testData = new RDDMiniBatches(batchSizePerWorker, pointsTest).miniBatchesJava();

        MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder()
                .seed(seed)
                .updater(new Nesterovs(learningRate, 0.9))
                .list()
                .layer(0, new DenseLayer.Builder().nIn(numInputs).nOut(numHiddenNodes)
                        .weightInit(WeightInit.XAVIER)
                        .activation(Activation.RELU)
                        .build())
                .layer(1, new OutputLayer.Builder(LossFunctions.LossFunction.NEGATIVELOGLIKELIHOOD)
                        .weightInit(WeightInit.XAVIER)
                        .activation(Activation.SOFTMAX)
                        .nIn(numHiddenNodes).nOut(numOutputs).build())
                .pretrain(false).backprop(true).build();


        TrainingMaster tm = new ParameterAveragingTrainingMaster.Builder(batchSizePerWorker)
                .averagingFrequency(5)
                .workerPrefetchNumBatches(2)
                .batchSizePerWorker(batchSizePerWorker)
                .build();

        //Create the Spark network
        SparkDl4jMultiLayer sparkNet = new SparkDl4jMultiLayer(sc, conf, tm);

        for ( int i = 0; i <numEpochs; i++) {
            sparkNet.fit(trainData);
            log.info("Completed Epoch {}", i);
            System.out.println("Completed Epoch : " + i);
        }

        System.out.println("Evaluate model....");
        Evaluation eval = sparkNet.doEvaluation(testData, batchSize, new Evaluation(numOutputs))[0];
        System.out.println(eval.stats());`

and DL4J version is below：

        <java.version>1.8</java.version>
        <nd4j.version>1.0.0-beta</nd4j.version>
        <dl4j.version>1.0.0-beta</dl4j.version>
        <datavec.version>1.0.0-beta</datavec.version>
        <nd4j.backend>nd4j-native-platform</nd4j.backend>
        
        <dl4j.spark.version>1.0.0-beta_spark_1</dl4j.spark.version>

**Comment by ustbxyls** [comment]
@AlexDBlack

**Comment by ustbxyls** [comment]
ok, here is my data , which is starun data in dl4j-examples, [https://github.com/deeplearning4j/dl4j-examples/tree/master/dl4j-examples/src/main/resources/classification](url)

code for read file:
`final String filenameTrain = new ClassPathResource("/classification/saturn_data_train.csv").getFile().getPath();
        final String filenameTest = new ClassPathResource("/classification/saturn_data_eval.csv").getFile().getPath();`
net parameters :
`Nd4j.ENFORCE_NUMERICAL_STABILITY = true;
        int batchSizePerWorker = 10;
        int numEpochs = 10;

        int batchSize = 50;
        int seed = 123;
        double learningRate = 0.005;

        int numInputs = 2;
        int numOutputs = 2;
        int numHiddenNodes = 20;`

and I do the count on stringTestData, 

` long testDataCount = stringTestData.count();
        System.out.println("stringTestData count is :  " + testDataCount);`

here is my result,
stringTestData count is :  100

=========================Confusion Matrix=========================
  0  1
-------
 26 22 | 0 = 0
  0 51 | 1 = 1

Confusion matrix format: Actual (rowClass) predicted as (columnClass) N times

**Comment by ustbxyls** [comment]
@AlexDBlack 

I will try your test case

**Comment by AlexDBlack** [comment]
OK, so this is due to RDDMiniBatches, which seems to have a bug.
I've fixed, but note that you don't even need it here - our spark implementation handles all the batching internally. So just remove the lines using RDDMiniBatches and you should be fine.

**Comment by AlexDBlack** [comment]
Essentially, yes. If you say batchSizePerWorker of 10, then it'll aggregate/merge DataSets so that they have 10 examples when performing training - regardles of how many examples are originally in the DataSets in the ```RDD<DataSet>```
